### PR TITLE
User Model

### PR DESCRIPTION
**Pull request by Sonali12920**
# Related Issue
-  issue #36 

# Proposed Changes
Created a User Model and Login-Signup Portal
Made all the changes asked by @aarushivij 

# Checklist
- [x] Clean code
- [x] Documentation
- [x] Proper Readme.md

# Screenshots
- Updated

Sign up 
Screenshot (561)

Sign in
Screenshot (562)

Received the same token

# Additional Info or Context (if any)

### CODE REVIEW

File: server/.env.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+  mongoUrl= "mongodb://AllNotes:<password>@cluster0-shard-00-00.oirkj.mongodb.net:27017,cluster0-shard-00-01.oirkj.mongodb.net:27017,cluster0-shard-00-02.oirkj.mongodb.net:27017/<dbname>?ssl=true&replicaSet=atlas-aq5o24-shard-0&authSource=admin&retryWrites=true&w=majority"
+  jwtkey= "Your JWTKEY goes here"
+
+
```
Comment (HarshCasper) [inline server/.env.example:5]: Remove the extra lines.

File: server/.env.example [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
```
Comment (HarshCasper) [inline server/.env.example:1]: Remove the extra line.

File: server/index.js [hunk 3]
```diff
@@ -1,18 +1,47 @@
 const express = require('express')
-const app = express();
+const bodyParser = require('body-parser')
+const mongoose = require('mongoose')
+const dotenv = require("dotenv");
+dotenv.config();
+
+const app = express()
 const port = 8000;
+// const { mongoUrl } = process.env.mongoUrl
```
Comment (HarshCasper) [inline server/index.js:9]: Remove the commented code.

File: server/index.js [hunk 4]
```diff
@@ -1,18 +1,47 @@
...
+mongoose.connection.on('error', (err) => {
+  console.log("Not Connected to mongo", err)
+})
+
+
+
+// app.post('/', (req, res) => {
+//   console.log(req.body)
+//   res.send('hello')
+// })
```
Comment (HarshCasper) [inline server/index.js:39]: Remove the extra lines and commented lines of Code

File: server/index.js [hunk 5]
```diff
@@ -1,18 +1,47 @@
...
+
+app.get('/', requireToken, (req, res) => {
+  res.send("Your email is " + req.user.email)
```
Comment (HarshCasper) [inline server/index.js:42]: Instead we can have a more User-Friendly message: `Welcome to the All-Notes API v1`

File: server/middleware/requireToken.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    jwt.verify(token, jwtkey, async (err, payload) => {
+        if (err) {
+            return res.status(401).send({ error: "You must be logged in" })
+        }
+        const { userId } = payload
+        const user = await User.findById(userId)
+        req.user = user
+        next();
+    })
+}
```
Comment (HarshCasper) [inline server/middleware/requireToken.js:23]: Add a new line at the EOF

File: server/models/User.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+//         };    
+
+//  //method to get the JSON representation of a user for authentication.
+//         UserSchema.methods.toAuthJSON = function(){
+//             return {
+//             username: this.username,
+//             email: this.email,
+//             token: this.generateJWT(),
+//             };
+//         };
```
Comment (HarshCasper) [inline server/models/User.js:79]: Remove the commented lines of Code

File: server/models/User.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+UserSchema.plugin(uniqueValidator, { message: 'is already taken.' });
+mongoose.model('User', UserSchema)
```
Comment (HarshCasper) [inline server/models/User.js:81]: Add a new line at the EOF

File: server/routes/authRoutes.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+const express = require('express')
+const mongoose = require('mongoose')
+const jwt = require('jsonwebtoken');
+const dotenv = require("dotenv");
+dotenv.config();
+// const { jwtkey } = process.env.jwtkey
```
Comment (HarshCasper) [inline server/routes/authRoutes.js:6]: Remove the commented line of Code

File: server/routes/authRoutes.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    res.send({ token })
+  }catch (err) {
+    return res.status(422).send({error:"You must provide email or password"})
+  }
+  
+
+})
+
+
+module.exports = router
```
Comment (HarshCasper) [inline server/routes/authRoutes.js:52]: Add a new line at the EOF

File: server/routes/authRoutes.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  if(!user){
+    return res.status(422).send({error:"You must provide email or password"})
+  }
+  try{
+    await user.comparePassword(password);
+    const token = jwt.sign({ userId: user._id }, process.env.jwtkey)
+    res.send({ token })
+  }catch (err) {
+    return res.status(422).send({error:"You must provide email or password"})
+  }
```
Comment (HarshCasper) [inline server/routes/authRoutes.js:46]: Format the Code

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Great work @Sonali12920 

However, I see that your commits are mixed with few other commits as well. Can you check why has that happened, and if possible revert those commits back?

**Comment by HarshCasper** [comment]
Hi @Sonali12920 

Are you still working on it?

**Comment by Sonali12920** [comment]
@HarshCasper, Yes I am still working on it.

**Pull request closed without merging**